frontend/packages/aquascope-editor/src/editor-utils/stepper.tsx

```diff
@@ -0,0 +1,383 @@
+          <i
+            className={`fa fa-${icon} aquascope-action-indicator`}
+            title={desc}
+          />
+        );
+        break loop;
+      }
+    }
+  }
+
+  let perms: [ValueStep<boolean>, string][] = [
+    [diffs.permissions.read, readChar],
+    [diffs.permissions.write, writeChar],
+    [diffs.permissions.drop, dropChar],
+  ];
+
+  let pathCol = <td className="perm-step-path">{path}</td>;
+
+  let permSpans = perms.map(([diff, content]) =>
+    diff.type == "High" ? (
+      <span key={content} className="perm-diff-add">
+        {content}
+      </span>
+    ) : diff.type == "Low" ? (
+      <span key={content} className="perm-diff-sub">
+        {content}
+      </span>
+    ) : diff.type === "None" && diff.value ? (
+      <span key={content} className="perm-diff-none-high">
+        {content}
+      </span>
+    ) : diff.type === "None" && !diff.value ? (
+      <span key={content} className="perm-diff-none-low">
+        -
+      </span>
+    ) : null
+  );
+
+  return (
+    <tr>
+      {pathCol}
+      <td>{ico}</td>
+      <td>{permSpans}</td>
+    </tr>
+  );
+};
+
+let stepLocation = (step: PermissionsStateStep): number => {
+  return step.location.char_end;
+};
+
+class PermissionStepTableWidget extends WidgetType {
+  pathsHidden: boolean;
+  tableHidden: boolean;
+  line: Line;
+  rowHTML?: string;
+  constructor(
+    readonly view: EditorView,
+    readonly config: StepperConfig,
+    readonly step: PermissionsStateStep,
+    readonly focusedLines: number[],
+    readonly focusedPaths: Map<number, RegExp>
+  ) {
+    super();
+    this.pathsHidden = true;
+    this.tableHidden = false;
+    this.line = view.state.doc.lineAt(stepLocation(step));
+  }
+
+  eq(other: PermissionStepTableWidget): boolean {
+    // Only one table widget can (should) be on a line.
+    return this.line.number === other.line.number;
+  }
+
+  toDOM() {
+    let container = document.createElement("span");
+
+    let doc = this.view.state.doc;
+    let pos = stepLocation(this.step);
+    let maxLineLen = 0;
+
+    for (let line of doc.iterLines()) {
+      maxLineLen = Math.max(maxLineLen, line.length);
+    }
+    let currLine = this.line;
+    let hideTable = !this.focusedLines.includes(currLine.number);
+    this.tableHidden = hideTable;
+
+    let padding = 2 + maxLineLen - currLine.length;
+    let spaces = " " + "_".repeat(padding);
+
+    let arrowOut = "»";
+    let arrowIn = "«";
+
+    let toggleHiddenPaths = (_e: any) => {
+      this.pathsHidden = !this.pathsHidden;
+      let content = container.querySelector<HTMLDivElement>(
+        ".step-table-content"
+      );
+      let dropdown = container.querySelector<HTMLDivElement>(
+        ".step-table-dropdown"
+      );
+      if (content) {
+        content.style.height = this.pathsHidden ? "0px" : "auto";
+      }
+      if (dropdown) {
+        dropdown.innerHTML = this.pathsHidden ? "..." : "";
+      }
+    };
+
+    let toggleWidget = (_e: any) => {
```

Comment:
This state should be within a React component, not on the Codemirror DOM widget. Eg

```tsx
let [tableHidden, setTableHidden] = useState(false);
return <button onClick={() => setTableHidden(!tableHidden)} />
```

It's also an anti-pattern to change styling directly in an event callback. I would only manipulate state on-click, and then leave the view function to decide how to render its contents given the current state.